File: app/web/src/api/sdf/dal/func.ts
import { ActionKind, ActionPrototypeId } from "@/api/sdf/dal/action";
import { InputSocketId, OutputSocketId, SchemaVariantId } from "./schema";
import { ComponentId } from "./component";
import { PropId } from "./prop";

export type FuncArgumentId = string;
export type FuncId = string;
export type AttributePrototypeArgumentId = string;
export type AttributePrototypeId = string;
export type ManagementPrototypeId = string;

export enum FuncKind {
  Action = "Action",
  Attribute = "Attribute",
  Authentication = "Authentication",
  CodeGeneration = "CodeGeneration",
  Intrinsic = "Intrinsic",
  Qualification = "Qualification",
  SchemaVariantDefinition = "SchemaVariantDefinition",
  Unknown = "Unknown",
  Management = "Management",
}

export enum CustomizableFuncKind {
  Action = "Action",
  Attribute = "Attribute",
  Authentication = "Authentication",
  CodeGeneration = "CodeGeneration",
  Qualification = "Qualification",
  Management = "Management",
}

// TODO(nick,wendy): this is ugly to use in some places. We probably need to think of a better interface. Blame me, not Wendy.
export function customizableFuncKindToFuncKind(customizableFuncKind: CustomizableFuncKind): FuncKind {
  switch (customizableFuncKind) {
    case CustomizableFuncKind.Action:
      return FuncKind.Action;
    case CustomizableFuncKind.Attribute:
      return FuncKind.Attribute;
    case CustomizableFuncKind.Authentication:
      return FuncKind.Authentication;
    case CustomizableFuncKind.CodeGeneration:
      return FuncKind.CodeGeneration;
    case CustomizableFuncKind.Qualification:
      return FuncKind.Qualification;
    case CustomizableFuncKind.Management:
      return FuncKind.Management;
    default:
      throw new Error("this should not be possible since CustomizableFuncKind is a subset of FuncKind");
  }
}

export type FUNC_LABELS = { pluralLabel: string; singularLabel: string };
export type FUNC_TYPES = Record<CustomizableFuncKind, FUNC_LABELS>;

export const CUSTOMIZABLE_FUNC_TYPES: FUNC_TYPES = {
  [CustomizableFuncKind.Action]: {
    pluralLabel: "Actions",
    singularLabel: "Action",
  },
  [CustomizableFuncKind.Attribute]: {
    pluralLabel: "Attributes",
    singularLabel: "Attribute",
  },
  [CustomizableFuncKind.Authentication]: {
    pluralLabel: "Authenticators",
    singularLabel: "Authentication",
  },
  [CustomizableFuncKind.CodeGeneration]: {
    pluralLabel: "Code Generators",
    singularLabel: "Code Generation",
  },
  [CustomizableFuncKind.Management]: {
    pluralLabel: "Management",
    singularLabel: "Management",
  },
  [CustomizableFuncKind.Qualification]: {
    pluralLabel: "Qualifications",
    singularLabel: "Qualification",
  },
};

export const isCustomizableFuncKind = (f: FuncKind) => f in CUSTOMIZABLE_FUNC_TYPES;

export enum FuncArgumentKind {
  Array = "array",
  Boolean = "boolean",
  Float = "float",
  Integer = "integer",
  Json = "json",
  Object = "object",
  String = "string",
  Map = "map",
  Any = "any",
}

export interface FuncArgument {
  id: string;
  name: string;
  kind: FuncArgumentKind;
  elementKind?: FuncArgumentKind;
  created_at: IsoDateString;
  updated_at: IsoDateString;
}

export enum FuncBackendKind {
  Array = "Array",
  Boolean = "Boolean",
  Diff = "Diff",
  Identity = "Identity",
  Float = "Float",
  Integer = "Integer",
  JsAction = "JsAction",
  JsAuthentication = "JsAuthentication",
  Json = "Json",
  JsReconciliation = "JsReconciliation",
  JsSchemaVariantDefinition = "JsSchemaVariantDefinition",
  Map = "Map",
  NormalizeToArray = "NormalizeToArray",
  Object = "Object",
  ResourcePayloadToValue = "ResourcePayloadToValue",
  String = "String",
  Unset = "Unset",
  Validation = "Validation",
}
export interface FuncSummary {
  funcId: FuncId;
  kind: FuncKind;
  name: string;
  displayName: string | null;
  description: string | null;
  isLocked: boolean;
  arguments: FuncArgument[];
  backendKind: FuncBackendKind;
  bindings: FuncBinding[];
  types?: string | null;
}

export interface FuncCode {
  funcId: FuncId;
  code: string;
}

export interface AttributeArgumentBinding {
  funcArgumentId: FuncArgumentId;
  attributePrototypeArgumentId: AttributePrototypeArgumentId | null;
  propId: PropId | null;
  inputSocketId: InputSocketId | null;
}

export enum FuncBindingKind {
  Action = "action",
  Attribute = "attribute",
  Authentication = "authentication",
  CodeGeneration = "codeGeneration",
  Qualification = "qualification",
  Management = "management",
}

export interface Action {
  bindingKind: FuncBindingKind.Action;
  // uneditable
  funcId: FuncId | null;
  schemaVariantId: SchemaVariantId | null;
  actionPrototypeId: ActionPrototypeId | null;
  // editable
  kind: ActionKind | null;
}

export interface Attribute {
  bindingKind: FuncBindingKind.Attribute;
  // uneditable
  funcId: FuncId | null;
  attributePrototypeId: AttributePrototypeId | null;
  // needed on create
  schemaVariantId: SchemaVariantId | null;
  componentId: ComponentId | null;
  // editable
  propId: PropId | null;
  outputSocketId: OutputSocketId | null;
  argumentBindings: AttributeArgumentBinding[];
}

export interface Management {
  bindingKind: FuncBindingKind.Management;
  // uneditable
  funcId: FuncId | null;
  managementPrototypeId: ManagementPrototypeId | null;
  // needed on create
  schemaVariantId: SchemaVariantId | null;
  componentId: ComponentId | null;
}

export interface Authentication {
  bindingKind: FuncBindingKind.Authentication;
  funcId: FuncId;
  schemaVariantId: SchemaVariantId;
}

export interface CodeGeneration {
  bindingKind: FuncBindingKind.CodeGeneration;
  funcId: FuncId | null;
  schemaVariantId: SchemaVariantId | null;
  attributePrototypeId: AttributePrototypeId | null;
  componentId: ComponentId | null;
  // editable
  inputs: LeafInputLocation[];
}

export interface Qualification {
  bindingKind: FuncBindingKind.Qualification;
  funcId: FuncId | null;
  schemaVariantId: SchemaVariantId | null;
  attributePrototypeId: AttributePrototypeId | null;
  componentId: ComponentId | null;
  // editable
  inputs: LeafInputLocation[];
}

export type FuncBinding = Action | Attribute | Authentication | CodeGeneration | Qualification | Management;

export type LeafInputLocation = "code" | "deletedAt" | "domain" | "resource" | "secrets";

export interface BindingWithBackendKind extends Attribute {
  backendKind: FuncBackendKind;
  attributePrototypeId: NonNullable<AttributePrototypeId>;
}

export interface PropDisplay {
  id: PropId;
  path: string;
  name: string;
  // eslint-disable-next-line @typescript-eslint/no-duplicate-type-constituents
  value?: PropId | InputSocketId;
  attributePrototypeId?: AttributePrototypeId;
  funcId: FuncId;
}

export interface BindingWithBackendKindAndPropId extends BindingWithBackendKind {
  propId: NonNullable<PropId>;
}

export interface IntrinsicDisplay {
  attributePrototypeId: AttributePrototypeId;
  outputSocketId: OutputSocketId;
  socketName: string;
  backendKind: FuncBackendKind;
  // eslint-disable-next-line @typescript-eslint/no-duplicate-type-constituents
  value: InputSocketId | PropId | undefined;
  funcId: FuncId;
}

// PSA: this is how to type guard filter so later operations know the field
// is no longer nullable b/c the filter removed any objects where the property was null
export interface BindingWithBackendKindAndOutputSocket extends BindingWithBackendKind {
  outputSocketId: NonNullable<OutputSocketId>;
}

File: lib/si-frontend-types-rs/src/func.rs
use serde::{
    Deserialize,
    Serialize,
};
pub use si_events::FuncKind;
use si_events::{
    ActionKind,
    ActionPrototypeId,
    AttributePrototypeArgumentId,
    AttributePrototypeId,
    ComponentId,
    FuncArgumentId,
    FuncBackendKind,
    FuncId,
    InputSocketId,
    ManagementPrototypeId,
    OutputSocketId,
    PropId,
    SchemaVariantId,
    Timestamp,
};
use si_id::{
    LeafPrototypeId,
    SchemaId,
};
use strum::{
    AsRefStr,
    Display,
    EnumIter,
    EnumString,
};

#[derive(Clone, Debug, Deserialize, Serialize, Eq, PartialEq)]
#[serde(rename_all = "camelCase")]
pub enum IntrinsicFuncKind {
    Identity,
    SetArray,
    SetBoolean,
    SetInteger,
    SetJson,
    SetMap,
    SetObject,
    SetString,
    Unset,
    Validation,
}

#[derive(Clone, Debug, Deserialize, Serialize, Eq, PartialEq)]
#[serde(rename_all = "camelCase")]
pub struct FuncSummary {
    pub func_id: FuncId,
    pub kind: FuncKind,
    pub name: String,
    pub display_name: Option<String>,
    pub description: Option<String>,
    pub is_locked: bool,
    pub arguments: Vec<FuncArgument>,
    pub bindings: Vec<FuncBinding>,
    pub types: Option<String>,
    pub backend_kind: FuncBackendKind,
    pub is_transformation: bool,
}

#[derive(Clone, Debug, Deserialize, Eq, Serialize, PartialEq)]
#[serde(rename_all = "camelCase")]
pub struct FuncArgument {
    pub id: Option<FuncArgumentId>,
    pub name: String,
    pub kind: FuncArgumentKind,
    pub element_kind: Option<FuncArgumentKind>,
    #[serde(flatten)]
    pub timestamp: Timestamp,
}

#[derive(Clone, Debug, Deserialize, Eq, Serialize, PartialEq)]
#[serde(rename_all = "camelCase")]
pub struct FuncCode {
    pub func_id: FuncId,
    pub code: String,
}
#[derive(Clone, Debug, Deserialize, Eq, Serialize, PartialEq)]
#[serde(rename_all = "camelCase")]
pub struct FuncBindings {
    pub bindings: Vec<FuncBinding>,
}
#[remain::sorted]
#[derive(
    AsRefStr,
    Clone,
    Debug,
    Deserialize,
    EnumString,
    Eq,
    Serialize,
    Display,
    EnumIter,
    PartialEq,
    Hash,
)]
#[serde(rename_all = "camelCase", tag = "bindingKind")]
pub enum FuncBinding {
    #[serde(rename_all = "camelCase")]
    Action {
        // unique ids
        schema_variant_id: Option<SchemaVariantId>,
        action_prototype_id: Option<ActionPrototypeId>,
        func_id: Option<FuncId>,
        //thing that can be updated
        kind: Option<ActionKind>,
    },
    #[serde(rename_all = "camelCase")]
    Attribute {
        // unique ids
        func_id: Option<FuncId>,
        attribute_prototype_id: Option<AttributePrototypeId>,
        // things needed for create
        component_id: Option<ComponentId>,
        schema_variant_id: Option<SchemaVariantId>,

        // things that can be updated
        prop_id: Option<PropId>,
        output_socket_id: Option<OutputSocketId>,

        // can optionally send arguments when creating the prototype,
        // or update them later individually
        argument_bindings: Vec<AttributeArgumentBinding>,
    },
    #[serde(rename_all = "camelCase")]
    Authentication {
        // unique ids
        schema_variant_id: SchemaVariantId,
        func_id: Option<FuncId>,
    },
    #[serde(rename_all = "camelCase")]
    CodeGeneration {
        // unique ids
        schema_variant_id: Option<SchemaVariantId>,
        component_id: Option<ComponentId>,
        func_id: Option<FuncId>,
        attribute_prototype_id: Option<AttributePrototypeId>,
        leaf_prototype_id: Option<LeafPrototypeId>,
        // thing that can be updated
        inputs: Vec<LeafInputLocation>,
    },
    #[serde(rename_all = "camelCase")]
    Management {
        schema_ids: Option<Vec<SchemaId>>,
        schema_variant_id: Option<SchemaVariantId>,
        management_prototype_id: Option<ManagementPrototypeId>,
        func_id: Option<FuncId>,
    },
    #[serde(rename_all = "camelCase")]
    Qualification {
        // unique ids
        schema_variant_id: Option<SchemaVariantId>,
        component_id: Option<ComponentId>,
        func_id: Option<FuncId>,
        attribute_prototype_id: Option<AttributePrototypeId>,
        leaf_prototype_id: Option<LeafPrototypeId>,
        // thing that can be updated
        inputs: Vec<LeafInputLocation>,
    },
}

impl FuncBinding {
    pub fn leaf_inputs(&self) -> Option<&[LeafInputLocation]> {
        match self {
            FuncBinding::CodeGeneration { inputs, .. }
            | FuncBinding::Qualification { inputs, .. } => Some(inputs),
            _ => None,
        }
    }

    pub fn leaf_binding_prototype(&self) -> Option<LeafBindingPrototype> {
        match self {
            FuncBinding::CodeGeneration {
                attribute_prototype_id,
                leaf_prototype_id,
                ..
            }
            | FuncBinding::Qualification {
                attribute_prototype_id,
                leaf_prototype_id,
                ..
            } => leaf_prototype_id
                .map(LeafBindingPrototype::Overlay)
                .or(attribute_prototype_id.map(LeafBindingPrototype::Attribute)),
            _ => None,
        }
    }

    pub fn schema_variant_id(&self) -> Option<SchemaVariantId> {
        match self {
            FuncBinding::Action {
                schema_variant_id, ..
            } => *schema_variant_id,
            FuncBinding::Attribute {
                schema_variant_id, ..
            } => *schema_variant_id,
            FuncBinding::Authentication {
                schema_variant_id, ..
            } => Some(*schema_variant_id),
            FuncBinding::CodeGeneration {
                schema_variant_id, ..
            } => *schema_variant_id,
            FuncBinding::Management {
                schema_variant_id, ..
            } => *schema_variant_id,
            FuncBinding::Qualification {
                schema_variant_id, ..
            } => *schema_variant_id,
        }
    }
}

#[derive(Deserialize, Serialize, Debug, Clone, Copy, PartialEq, Eq, Hash)]
pub enum LeafBindingPrototype {
    Attribute(AttributePrototypeId),
    Overlay(LeafPrototypeId),
}

impl From<LeafBindingPrototype> for Option<AttributePrototypeId> {
    fn from(value: LeafBindingPrototype) -> Self {
        match value {
            LeafBindingPrototype::Attribute(attribute_prototype_id) => Some(attribute_prototype_id),
            LeafBindingPrototype::Overlay(_) => None,
        }
    }
}

impl From<LeafBindingPrototype> for Option<LeafPrototypeId> {
    fn from(value: LeafBindingPrototype) -> Self {
        match value {
            LeafBindingPrototype::Attribute(_) => None,
            LeafBindingPrototype::Overlay(leaf_prototype_id) => Some(leaf_prototype_id),
        }
    }
}

#[derive(Clone, Debug, Deserialize, Eq, Serialize, PartialEq, Hash)]
#[serde(rename_all = "camelCase")]
pub struct AttributeArgumentBinding {
    pub func_argument_id: FuncArgumentId,
    pub attribute_prototype_argument_id: Option<AttributePrototypeArgumentId>,
    pub prop_id: Option<PropId>,
    pub input_socket_id: Option<InputSocketId>,
    pub static_value: Option<serde_json::Value>,
}

/// This enum provides available child [`Prop`](crate::Prop) trees of [`RootProp`](crate::RootProp)
/// that can be used as "inputs" for [`Funcs`](crate::Func) on leaves.
///
/// _Note: not all [`children`](crate::RootPropChild) of [`RootProp`](crate::RootProp) can be used
/// as "inputs" in order to prevent cycles. This enum provides an approved subset of those
/// children_.
#[remain::sorted]
#[derive(
    AsRefStr,
    Clone,
    Copy,
    Debug,
    Deserialize,
    EnumString,
    Eq,
    Serialize,
    Display,
    EnumIter,
    Ord,
    PartialEq,
    PartialOrd,
    Hash,
)]
#[serde(rename_all = "camelCase")]
pub enum LeafInputLocation {
    /// The input location corresponding to "/root/code".
    Code,
    /// The input location corresponding to "/root/deleted_at"
    DeletedAt,
    /// The input location corresponding to "/root/domain".
    Domain,
    /// The input location corresponding to "/root/resource".
    Resource,
    /// The input location corresponding to "/root/secrets".
    Secrets,
}

#[remain::sorted]
#[derive(
    AsRefStr,
    Clone,
    Copy,
    Debug,
    Deserialize,
    EnumString,
    Eq,
    Serialize,
    Display,
    EnumIter,
    Ord,
    PartialEq,
    PartialOrd,
    Hash,
)]
#[serde(rename_all = "camelCase")]
pub enum FuncArgumentKind {
    Any,
    Array,
    Boolean,
    Float,
    Integer,
    Json,
    Map,
    Object,
    String,
}
